### ISSUE TITLE
Add linting to CI

### ISSUE THREAD

**Issue by rylnd** [rylnd/shpec#60]
This has been spiked out in #66, using [shellcheck](http://www.shellcheck.net/about.html), but we should discuss exactly how (and why) we would want this functionality.

**Comment by locochris** [comment]
(At the moment) its an expensive test/check to run (ie. over `bin/shpec`), but I was just interested to have a look and see if anyone else thought it might be useful.

**Comment by rylnd** [comment]
Yeah, I think initially it'd be useful to make sure we're not doing anything dumb. If we can use shellcheck  to automatically enforce style on a PR, that seems like it would be nice.

Regarding speed, we could probably just run our builds on the haskell VM, which already has cabal installed. We should also be able to use Travis' [container-based infrastructure](http://docs.travis-ci.com/user/workers/container-based-infrastructure/) to do some directory-based caching (of shellcheck).

So, I think the answer to "Can we?" is "Yes," but the "Should we?" remains to be seen. Without sufficient configuration (e.g. disabling a check that you disagree with), I can imagine such a tool becoming reallll annoying. Does shellcheck (or any similar tool) provide such configuration?

**Comment by locochris** [comment]
I'm happy to blindly follow the herd's strong opinions on this one - especially when they're built up through lots of experience with all the shell idiosyncrasies.  I suppose you could judge the impartialness of this linter based on the acceptance or rejection of all its pull requests, but I reckon just go with it.

BTW have you tried setting it up with syntastic in vim?
![screen shot 2015-04-22 at 10 25 01 am](https://cloud.githubusercontent.com/assets/2774052/7265250/fe38bd30-e8d9-11e4-82d8-2c27382514eb.png)

**Comment by rylnd** [comment]
Okay, so I tried [a bunch of configurations](https://travis-ci.org/rylnd/shpec/builds/), here's what I found:

1. Installing cabal and shellcheck on the default VM is surprisingly faster than installing shellcheck on the haskell VM. Compare [default](https://travis-ci.org/rylnd/shpec/builds/59797824) to [haskell](https://travis-ci.org/rylnd/shpec/builds/59811908)
2. Running an OSX VM and using homebrew to install everything [runs the quickest](https://travis-ci.org/rylnd/shpec/builds/59823886), but the boot times are incredibly (read: prohibitively) slow.
3. I was unable to get any cabal caching to work. I was able to cache and restore the artifacts of `cabal install shellcheck`, but cabal just ignored them and reinstalled everything. Currently, even if we were able to make it work, it would prevent the use of sudo which makes installing other dependencies tricky. There seems to be some precompiled shellcheck packages out there that don't require cabal, and [Travis' docs](http://docs.travis-ci.com/user/installing-dependencies/) might be of assistance, but I personally lost patience. Maybe someone with better linux chops can figure it out.

In short, installing shellcheck takes way too long. I was unable to get it cached between builds. If we _are_ able to cache shellcheck, we then have to figure out how to install other shells (`ksh`, `zsh`) without sudo.

**Comment by locochris** [comment]
@rylnd - its fiddly isn't it.
There's "A submodule to lint your shell projects with shellcheck in travis.ci builds" https://github.com/caarlos0/shell-travis-build
This guy is installing shellcheck using https://github.com/caarlos0/shell-travis-build/blob/master/install.sh#L5-L7

I gave it a go in https://github.com/rylnd/shpec/pull/66 (https://github.com/locochris/shpec/commit/caafaea5024886360b5990da341b3e9bea420f2a) and the build times aren't too shabby.  ([3s to install shellcheck](https://travis-ci.org/rylnd/shpec/builds/59831385))

Perhaps this is acceptable? (All our time is going to go into installing shells anyway - and we need `sudo` to do that - so we'll have to do all this without caching anyway.)

Then I gave it another go in (https://github.com/rylnd/shpec/commit/97079fca197a08c63e56217ab2d53ca6ced779c3) which only bothers to run the linter using the `sh` dialect. (see [travis build](https://travis-ci.org/rylnd/shpec/builds/59832914))

**Comment by rylnd** [comment]
@locochris brilliant, installing the `.deb` is exactly what I wanted to do, I just couldn't find a URL. I think that 3s is totally acceptable. We're closing this in favor of #68, then?
